feat: add end to end tests for features in SAM

ShreyaGangishetty opened the pull request:
*Issue #, if available:*

*Description of changes:*
Added basic setup for being able to write end to end tests for SAM

*Description of how you validated changes:*
Ran the end to end tests using `make e2e-tests`

*Checklist:*

- [x] Write/update tests
- [x] `make pr` passes
- [x] Update documentation
- [ ] Verify transformed template deploys and application functions as expected
- [ ] Add/update example to `examples/2016-10-31`


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

jfuss commented on end-to-end-tests/tst/helpers/base_test.py:
```diff
@@ -0,0 +1,86 @@
+    def get_sam_transformed_template(self, template_path):
+        try:
+            subprocess.call(
+                ["bin/sam-translate.py", "--template-file", template_path, "--o", self.stack_name + ".json"]
```
Why are we shelling out here? Can't we just run the transform?

ShreyaGangishetty replied:
I could not import sam-translate even after adding `__init__.py` to bin folder. hence i used subprocess

ShreyaGangishetty replied:
I will update in the next revision to use transform code

jfuss commented on end-to-end-tests/tst/helpers/base_test.py:
```diff
@@ -0,0 +1,86 @@
+
+class BaseTest(unittest.TestCase):
+    def setUp(self):
+        self.stack_name = "sam-temp-stack" + str(random.randint(0, 10000))
```
instead of `random.randint` you should use `uuid`

ShreyaGangishetty replied:
thank you, updated the code with `uuid`

jfuss commented on end-to-end-tests/tst/helpers/resource_status.py:
```diff
@@ -0,0 +1,22 @@
+from enum import Enum
+
+# contains all the the cloudformation stack resource status
+class ResourceStatus(Enum):
```
What's the value of this being an Enum here?

ShreyaGangishetty replied:
To state that the ResourceStatus is a constant so I used Enum

praneetap commented on end-to-end-tests/tst/helpers/base_test.py:
```diff
@@ -0,0 +1,86 @@
+        self.stack_name = "sam-temp-stack" + str(random.randint(0, 10000))
+        self.cloud_formation_client = boto3.client("cloudformation")
+
+    def make_stack(self, capabilties, template_path):
```
SHouldn't we raise an exception if this fails

ShreyaGangishetty replied:
Updated the code with exceptions where ever applicable

praneetap commented on end-to-end-tests/tst/helpers/base_test.py:
```diff
@@ -0,0 +1,86 @@
+                outputs[output.get("OutputKey")] = output.get("OutputValue")
+            return outputs
+
+    def delete_stack(self):
```
raise an exception if delete fails

ShreyaGangishetty replied:
Cloudformation doesn't throw an exception for delete stack. https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/cloudformation.html#CloudFormation.Client.delete_stack
I will try to add additional check to throw exception when delete fails

jfuss commented on end-to-end-tests/tst/helpers/resources.py:
```diff
@@ -0,0 +1,129 @@
+        if not isinstance(other, Resources):
+            return False
+        # checking if the keys are same
+        if self.resources.keys() != other.resources.keys():
```
Can't we just compare the dictionaries? ` self.resources == other.resources`?

ShreyaGangishetty replied:
Thank you, updated the code to just compare the dictionaries directly and it works as expected

jfuss commented on end-to-end-tests/tst/helpers/resources.py:
```diff
@@ -0,0 +1,129 @@
+        return True
+
+    """
+    Validates and Cleans up LogicalIDs generated by SAM translator.
```
Random doc strings? Can this be placed right under the function/class/etc definitions?

ShreyaGangishetty replied:
Thank you, I have updated the code with doc strings for class and method definitions

jfuss commented on end-to-end-tests/tst/helpers/resources.py:
```diff
@@ -0,0 +1,129 @@
+class Resource:
+    # We only care about the following properties when comparing for equality because rest are dynamically
+    # generated by CloudFormation: LogicalResourceId, ResourceStatus, ResourceType
+    logical_id = str()
```
Why are these defined globally on Resource?

ShreyaGangishetty replied:
Good catch! We don't need these attributes to be present globally. I have updated the code

jfuss commented on end-to-end-tests/tst/helpers/resources.py:
```diff
@@ -0,0 +1,129 @@
+import re
+
+
+class Resources:
```
Can we add docstrings to all public functions? It will make it easier to understand everything.

ShreyaGangishetty replied:
Good suggestion! added docstrings to all functions and classes

jfuss commented on end-to-end-tests/tst/test_http_api_cors.py:
```diff
@@ -0,0 +1,47 @@
+    def test_cors(self):
+        input_template = os.path.join(os.getcwd(), "end-to-end-tests/input", "http_api_cors", "template.yaml")
+        capabilities = ["CAPABILITY_IAM"]
+        stack_name = self.make_and_verify_stack(input_template, capabilities, self.http_api_with_cors_resources)
```
Can we split the make and verify steps into separate functions? It will make it easier to reason about when each step does one operation.

ShreyaGangishetty replied:
Agreed. I have added three functions: `make_stack`, `verify_stack` and `make_and_verify_stack`. We can either use make step then verify or use a single function `make_and_verify_stack` now

jfuss commented on end-to-end-tests/tst/test_rest_api_with_authorizers.py:
```diff
@@ -0,0 +1,131 @@
+            lambda_authorizer_uri,
+            "lambdaTokenAuthorizer: authorizer URI must be the Lambda Function Authorizer's URI",
+        )
+        self.assertEquals(
```
`self.assertEquals` should be `self.assertEqual` (this is how it is done in python3).

praneetap replied:
+1. This should come up as a warning in build, please check if there are any other warnings as well.

ShreyaGangishetty replied:
Initially I had written the code in Python2. Updated the code to work in python3

jfuss commented:
What Python version are we running these tests in? Can we make it so they only run in Python3? It will make our lives much much easier to not port more code into Python3 after Python2 finally dies.

jfuss commented on DEVELOPMENT_GUIDE.rst:
```diff
@@ -70,6 +70,13 @@ drops below 95%.
+
 Pull Requests
 -------------
 Before sending pull requests make sure to run ``make pr`` command. This will run unit tests, linters, and static
```
Do we still only want `make pr` to work? Should running integ tests be apart of `make pr`?

ShreyaGangishetty replied:
I have added [here](https://github.com/awslabs/serverless-application-model/pull/1557/files#diff-195a635ad245ded9076330e31134bd80R12) but forgot to add in this page. I will update it
I didn't add a way to running integ tests as part of build but we will eventually want integ tests to be a part of make pr.

ShreyaGangishetty commented on DEVELOPMENT_GUIDE.rst:
```diff
@@ -70,6 +70,13 @@ drops below 95%.
+End to end tests
+~~~~~~~~~~~~~~~~
+
+``make e2e tests`` command will run all end to end tests. This command is configured to fail when code coverage for package drops below 95%.
```
agreed, removed the code coverage check in the next revision

ShreyaGangishetty commented:
> What Python version are we running these tests in? Can we make it so they only run in Python3? It will make our lives much much easier to not port more code into Python3 after Python2 finally dies.

Initially I was running the tests in Python2. I have updated the code to run tests in Python3

jfuss commented on end_to_end_tests/input/rest_api_with_authorizers/template.yaml:
```diff
@@ -0,0 +1,134 @@
+    Properties:
+      Handler: index.handler
+      Runtime: nodejs12.x
+      InlineCode: |
```
Should one of these tests not use inline so we can see how that is setup? I would assume that InlineCode is not going to be the majority of things we do with these tests.

ShreyaGangishetty replied:
Good idea! I will update one of the tests without `InlineCode`

ShreyaGangishetty replied:
Updated the `template.yaml` for `http_api_cors`integration test

jfuss commented on end_to_end_tests/test/helpers/resources.py:
```diff
@@ -0,0 +1,247 @@
+            # out this suffix here. Tests can continue to access and verify the resource using the LogicalId
+            # specified in the template
+            if stack_resource.get("LogicalResourceId") is None:
+                raise ValueError("LogicalResourceId should not be empty for stack resource {}".format(stack_resource))
```
specific errors

jfuss commented on end_to_end_tests/test/test_rest_api_with_authorizers.py:
```diff
@@ -0,0 +1,199 @@
+from .helpers.base_test import BaseTest
+from .helpers.resources import Resources
+from .helpers.resource_types import ResourceTypes
+import urllib3
```
Any reason we choice urllib3 over requests?

ShreyaGangishetty replied:
No specific reason. I will update the code to use `requests` module

jfuss commented on pytest.ini:
```diff
@@ -1,7 +1,4 @@
 [pytest]
-# Fail if coverage falls below 95%
-# NOTE: If debug breakpoints aren't working, comment out the code coverage line below
-addopts = --cov samtranslator --cov-report term-missing --cov-fail-under 95
```
Is the coverage in all the commands we need it to be on the makefile?

ShreyaGangishetty replied:
Yes, right now we have only `make pr` which needs coverage and --cov is present in the Makefile already for `make pr`. Currently, we don't have 95% test coverage for integ tests. Removing these lines was necessary for `make e2e-tests` to pass

jfuss commented on requirements/dev.txt:
```diff
@@ -10,6 +10,10 @@ pytest>=3.0.7
 
+# Requirements for integration tests
+certifi>=2020.4.5.1
+urllib3
```
Wait, we already have a dependency on requests but add urllib3? We should just be able to use requests for calls.

ShreyaGangishetty replied:
Updated the code to use `requests`

ShreyaGangishetty commented on end_to_end_tests/test/helpers/base_test.py:
```diff
@@ -0,0 +1,199 @@
+            output_file = self.stack_name + ".json"
+            with open(output_file, "w") as fp:
+                pass
+            subprocess.call(["python", "bin/sam-translate.py", "--template-file", template_path, "--o", output_file])
```
I was not able to import the function even though i added `__init__.py` for bin to make that as a module. Hence, kept subprocess

ShreyaGangishetty commented on end_to_end_tests/test/helpers/base_test.py:
```diff
@@ -0,0 +1,199 @@
+        else:
+            if os.path.exists(self.stack_name + ".json"):
+                template_body = open(self.stack_name + ".json").read()
+                # delete the temporary file created
```
Thank you, I am removing `subprocess` to call the transform function, So I wont be needing to create temp files anymore.

ShreyaGangishetty commented on end_to_end_tests/test/helpers/base_test.py:
```diff
@@ -0,0 +1,199 @@
+from samtranslator.model.exceptions import InvalidTemplateException
+
+
+class BaseTest(unittest.TestCase):
```
Thanks you. I will update in the next revision

ShreyaGangishetty commented on end_to_end_tests/test/helpers/base_test.py:
```diff
@@ -0,0 +1,199 @@
+            waiter.wait(StackName=self.stack_name)
+            return stack_id
+        except Exception as e:
+            raise Exception("Unable to create stack for template {}. Exception: ".format(template_path, e))
```
removed the generic exception here and not catching any other exception as the error from CloudFormation would be self explanatory

ShreyaGangishetty commented:
> We need to be very careful about exceptions. We should not catch broad ones (if ever). So things like `except Exception` need to be scoped. This can lead to hiding actually errors which will give the tests a false sense of security.
> 
> We should avoid raising ValueError or other generic exceptions as well. If python ends up throwing one of this, we think it is one error but it is a completely different one. This can be very dangerous for integration tests that we rely on for deployments.

Thank you for the suggestion, it was really useful. Removed all the generic Exceptions and updated with user defined exceptions where ever applicable.
